# 4.5 Data Consistency Checks

## Table of Contents

#### Importing Libraries
#### Data Sets
#### Task

## Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Data Sets

In [2]:
# importing data sets
path=r'/Users/isabelgonzalez/Desktop/achievement 4'

In [56]:
df_ords = pd.read_csv(os.path.join(path, 'data', 'prepared data', 'orders_wrangled.csv'), index_col = False)

In [4]:
df_prods=pd.read_csv(os.path.join(path, 'data', 'original data', 'products.csv'), index_col = False)

## Task

In [25]:
#Step 1
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
#Adressing missing values
df_prods.shape

(49693, 5)

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

In [26]:
#Step 2
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max under the prices column is 99999 which seems way too high, it might be a mistake.

In [57]:
#Step 3
#Checking for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [34]:
#Step 4 
There are no mixed data types

In [58]:
#Step 5
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
customer_id                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in days_since_prior_order column.

In [59]:
# step 6
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [60]:
df_nan_ords

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  \
0                 0   2539329            1             1                   2   
11               11   2168274            2             1                   2   
26               26   1374495            3             1                   1   
39               39   3343014            4             1                   6   
45               45   2717275            5             1                   3   
...             ...       ...          ...           ...                 ...   
3420930     3420930    969311       206205             1                   4   
3420934     3420934   3189322       206206             1                   3   
3421002     3421002   2166133       206207             1                   6   
3421019     3421019   2227043       206208             1                   1   
3421069     3421069   3154581       206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

206209 is a very high amount to just replace or delete all values.
The reason these cells are probably empty is because there are that many customers who have only had one order, therefore, there are no days since a prior order. It would be best to create a new variable to determine one time customers and returning customers.

In [61]:
df_ords_clean=df_ords

In [62]:
df_ords_clean['one_time_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [63]:
df_ords_clean

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  \
0                 0   2539329            1             1                   2   
1                 1   2398795            1             2                   3   
2                 2    473747            1             3                   3   
3                 3   2254736            1             4                   4   
4                 4    431534            1             5                   4   
...             ...       ...          ...           ...                 ...   
3421078     3421078   2266710       206209            10                   5   
3421079     3421079   1854736       206209            11                   4   
3421080     3421080    626363       206209            12                   1   
3421081     3421081   2977660       206209            13                   1   
3421082     3421082    272231       206209            14                   6   

         order_hour_of_day  days_since_prior_order  one_time_customer  
0                        8                     NaN               True  
1                        7                    15.0              False  
2                       12                    21.0              False  
3                        7                    29.0              False  
4                       15                    28.0              False  
...                    ...                     ...                ...  
3421078                 18                    29.0              False  
3421079                 10                    30.0              False  
3421080                 12                    18.0              False  
3421081                 12                     7.0              False  
3421082                 14                    30.0              False  

[3421083 rows x 8 columns]

In [64]:
#step 7 checking for duplicates
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [65]:
# step 8
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, customer_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, one_time_customer]
Index: []

There are no duplicates.

In [66]:
# step 9 exporting data
df_ords_clean.to_csv(os.path.join(path, 'data','prepared Data', 'orders_clean.csv'))

In [54]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'data','prepared Data', 'products_clean.csv'))